rust/kernel/io_mem.rs

```diff
@@ -0,0 +1,195 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! Memory-mapped IO.
+//!
+//! C header: [`include/asm-generic/io.h`](../../../../include/asm-generic/io.h)
+
+use crate::{bindings, c_types, Error, Result};
+use core::convert::TryInto;
+
+extern "C" {
+    fn rust_helper_ioremap(
+        offset: bindings::resource_size_t,
+        size: c_types::c_ulong,
+    ) -> *mut c_types::c_void;
+
+    fn rust_helper_readb(addr: *const c_types::c_void) -> u8;
+    fn rust_helper_readw(addr: *const c_types::c_void) -> u16;
+    fn rust_helper_readl(addr: *const c_types::c_void) -> u32;
+    fn rust_helper_readq(addr: *const c_types::c_void) -> u64;
+
+    fn rust_helper_writeb(value: u8, addr: *mut c_types::c_void);
+    fn rust_helper_writew(value: u16, addr: *mut c_types::c_void);
+    fn rust_helper_writel(value: u32, addr: *mut c_types::c_void);
+    fn rust_helper_writeq(value: u64, addr: *mut c_types::c_void);
+}
+
+/// Represents a memory resource.
+pub struct Resource {
+    offset: bindings::resource_size_t,
+    size: bindings::resource_size_t,
+}
+
+impl Resource {
+    pub(crate) fn new(
```

Comment:
This should be `unsafe`, right? `IoMem` depends on the fact that you can't duplicate a `Resource`.

Reply:
I don't think `IoMem` depends on this for safety. Having the same range mapped multiple times should be ok; it may be bad for correctness, but why would it violate safety? We never use the returned pointer directly in Rust, only as arguments to readX/writeX.

I'm more concerned about the DMA case. My thinking was to make the DMA interface unsafe, but now that Gary mentioned it, it occurs to me that even a DMA "handle" is just a number that could be written to a device. IOW, what forces a driver to go through the unsafe DMA interface when it could just encode some arbitrary physical address and ask a bus mastering device to read/write it? Nothing at the moment, so this needs more thought.